File: oracle/js/src/priceSources/steemPrices.ts
// priceSources/steemPrices.ts — prices STEEM/SBD_Internal (internal market)
// and Price_Feed (Steem's own witness-median feed price) from the same
// Steem RPC node used for bridge scanning. Mirrors the Steem-sourced half of
// oracle/go/relayer/pricesource.go's CompositePriceSource dispatch (see
// oracle/PROTOCOL.md §7's price source → pair mapping table). Price_Feed is
// deliberately NOT pair-shaped — it's a single blockchain-native value, not
// a tradeable market rate.

import type { SteemClient } from "../steemClient";

export class SteemPriceSource {
  constructor(private readonly steem: SteemClient) {}

  /** Prices whichever of STEEM/SBD_Internal / Price_Feed are in `pairs`.
   * Never throws: a fetch failure for one pair just omits it from the
   * result, matching the "partial/empty map is fine" contract every
   * PriceSource shares (see priceFeeder.ts). */
  async fetchPrices(pairs: string[]): Promise<Record<string, string>> {
    const want = new Set(pairs);
    const out: Record<string, string> = {};

    if (want.has("STEEM/SBD_Internal")) {
      try {
        out["STEEM/SBD_Internal"] = await this.steem.getTicker();
      } catch {
        // omit — see doc comment above
      }
    }
    if (want.has("Price_Feed")) {
      try {
        out["Price_Feed"] = await this.steem.getFeedHistory();
      } catch {
        // omit — see doc comment above
      }
    }
    return out;
  }
}

File: oracle/python/src/steemvm_oracle/pricesources/steem_prices.py
"""STEEM/SBD_Internal (internal market) and Price_Feed (Steem's own
witness-median feed price) pricing, both sourced from the SAME Steem RPC
node used for bridge scanning -- no separate endpoint needed. See
``oracle/PROTOCOL.md`` SS7's price-source table and
``oracle/go/relayer/steem.go``'s ``GetTicker``/``GetFeedHistory`` (this
module is a thin per-pair wrapper around those, mirroring the Go client's
``CompositePriceSource`` dispatch for just these two pairs). Price_Feed is
deliberately NOT pair-shaped -- it's a single blockchain-native value, not
a tradeable market rate.
"""

from __future__ import annotations

from decimal import Decimal

from ..steem_client import SteemClient

PAIR_STEEM_SBD_INTERNAL = "STEEM/SBD_Internal"
PAIR_PRICE_FEED = "Price_Feed"


class SteemPriceSource:
    def __init__(self, steem: SteemClient):
        self._steem = steem

    def fetch_prices(self, pairs: list[str]) -> dict[str, Decimal]:
        """Prices whichever of STEEM/SBD_Internal, Price_Feed are requested.
        Never raises: a failure for one pair just omits it, matching the
        "partial/empty map is fine" contract every price source shares."""
        out: dict[str, Decimal] = {}
        if PAIR_STEEM_SBD_INTERNAL in pairs:
            try:
                out[PAIR_STEEM_SBD_INTERNAL] = self._steem.get_ticker()
            except Exception:  # noqa: BLE001 - a flaky Steem RPC just skips this pair
                pass
        if PAIR_PRICE_FEED in pairs:
            try:
                out[PAIR_PRICE_FEED] = self._steem.get_feed_history()
            except Exception:  # noqa: BLE001
                pass
        return out
